File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static ArrayList<Integer>[] list;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        list = new ArrayList[n];
        
        for(int i=0; i<n; i++){
            list[i] = new ArrayList<>();
        }
        
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(computers[i][j] == 1){
                    list[i].add(j);
                    list[j].add(i);
                }
            }
        }
        
        boolean[] visited = new boolean[n];
        for(int i=0; i<n; i++){
            if(!visited[i]){
                answer++;
                dfs(visited, i);
            }
        }
        
        return answer;
    }
    private static void dfs(boolean[] visited, int x){
        visited[x] = true;
        
        for(int i : list[x]){
            if(!visited[i])
                dfs(visited, i);
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    visited = [False] * (n)
    
    def dfs(node):
        visited[node] = True
        
        for next_node in range(n):
            # 연결되어 있고, 아직 방문하지 않았다면 방문
            if computers[node][next_node] == 1 and not visited[next_node]:
                dfs(next_node)
    
    for i in range(n):
        if not visited[i]:
            answer += 1
            dfs(i)
    
    
    return answer